src/http/ws_builder.h

```diff
@@ -0,0 +1,292 @@
+  }
+
+  class Message
+  {
+  protected:
+    HeaderMap headers;
+    const uint8_t* body = nullptr;
+    size_t body_size = 0;
+
+    Message() = default;
+
+  public:
+    const HeaderMap& get_headers() const
+    {
+      return headers;
+    }
+
+    void set_header(std::string k, const std::string& v)
+    {
+      // Store all headers lower-cased to simplify case-insensitive lookup
+      std::transform(k.begin(), k.end(), k.begin(), [](unsigned char c) {
+        return std::tolower(c);
+      });
+      headers[k] = v;
+    }
+
+    void clear_headers()
+    {
+      headers.clear();
+    }
+
+    size_t get_content_length() const
+    {
+      if (body == nullptr)
+      {
+        return 0;
+      }
+      else
+      {
+        return body_size;
+      }
+    }
+
+    const uint8_t* get_content_data() const
+    {
+      return body;
+    }
+
+    void set_body(const std::vector<uint8_t>* b)
+    {
+      if (b != nullptr)
+      {
+        set_body(b->data(), b->size());
+      }
+      else
+      {
+        set_body(nullptr, 0);
+      }
+    }
+
+    void set_body(const uint8_t* b, size_t s)
+    {
+      body = b;
+      body_size = s;
+
+      headers[headers::CONTENT_LENGTH] =
+        fmt::format("{}", get_content_length());
+    }
+  };
+
+  class Request : public Message
+  {
+  private:
+    http_method method;
+    std::string path = "/";
+    std::map<std::string, std::string> query_params = {};
+
+  public:
+    Request(const std::string_view& p = "/", http_method m = HTTP_POST) :
+      Message(),
+      method(m)
+    {
+      set_path(p);
+    }
+
+    http_method get_method() const
+    {
+      return method;
+    }
+
+    void set_path(const std::string_view& p)
+    {
+      if (p.size() > 0 && p[0] == '/')
+      {
+        path = p;
+      }
+      else
+      {
+        path = fmt::format("/{}", p);
+      }
+    }
+
+    std::string get_path() const
+    {
+      return path;
+    }
+
+    void set_query_param(const std::string& k, const std::string& v)
+    {
+      query_params[k] = v;
+    }
+
+    void set_query_params(const nlohmann::json& j)
+    {
+      for (auto it = j.begin(); it != j.end(); ++it)
+      {
+        set_query_param(it.key(), it.value().dump());
+      }
+    }
+
+    std::string get_formatted_query() const
+    {
+      std::string formatted_query;
+      bool first = true;
+      for (const auto& it : query_params)
+      {
+        formatted_query +=
+          fmt::format("{}{}={}", (first ? '?' : '&'), it.first, it.second);
+        first = false;
+      }
+      return formatted_query;
+    }
+
+    std::vector<uint8_t> build_request(bool header_only = false) const
+    {
+      const auto uri = fmt::format("{}{}", path, get_formatted_query());
+
+      const auto body_view = (header_only || body == nullptr) ?
+        std::string_view() :
+        std::string_view((char const*)body, body_size);
+
+      const auto request_string = fmt::format(
+        "{} {} HTTP/1.1\r\n"
```

Comment:
A lot of this file seems to be HTTP-specific. What does WebSockets use it for?

Reply:
That's unused now, I've removed it.